### DISCUSSION TITLE
Indexing errors during L-BFGS relaxation

### DISCUSSION THREAD

**Discussion by lionelalexandre** [OrderN/CONQUEST-release#109] [General]
Dear All,

this error message *may* appear when doing geometry optimisation using lbfgs
for molecule in a box. It doesn't seem to appear for 'cg' or 'sqnm' ; not tested with MD.

Inputs files are provided
[Conquest_halo_rem_error.tar.gz](https://github.com/OrderN/CONQUEST-release/files/8771628/Conquest_halo_rem_error.tar.gz)

Best regards,
Lionel

**Comment by davidbowler** [comment]
When I run this on two processes, I get the error:

index_trans: i_h2d wrong          0      3157

Is that what you see, or do you get the error in the title?

**Comment by lionelalexandre** [comment]
I get the error in the title ; using a different compilor (intel), I obtain something close to your error.

**Reply by davidbowler** [reply]
This is very interesting! I used GCC 11.2 and OpenMPI 4.1.3.  I will check Intel as well.  Which compiler were you using? Getting different errors with different compilers is rather strange.

**Comment by davidbowler** [comment]
I get the same error (index_trans: i_h2d wrong) with Intel 2016 and PGI 18.5

What compilers give you the halo_rem error?

**Comment by lionelalexandre** [comment]
gcc7

**Reply by davidbowler** [reply]
OK - gcc 4.8.5 (see below) it runs without any error; gcc 9 and 11 it stops with i_h2d error.

**Comment by davidbowler** [comment]
With gcc4.8.5 it runs without the error! This is *very* strange.

I am using Conquest version: 

develop; tag, hash: v1.0.5-pre-114-ge0f11dc

**Comment by negf** [comment]
I got the same error (index_trans: i_h2d wrong          0      3157) with intel 2021. i did run it through valgrind, and got an uninitialized value for g0 in subroutine lbfgs (control.f90) somewhere in the do while loop at "else if(abs(gg/g0)>ten) then". Initializing g0 before the do while loop solved it. But, not sure if it is related, after some iteration I get another error "Atoms are too close to each other:        10        12" and the structure looks very strange. Maybe I did initialize g0 in a strange way.

**Reply by davidbowler** [reply]
Thank you, Marius - that was the cause of the i_h2d error (after fixing it, I find no differences between one process and two processes, which was the problem before).  The later error seems to come from the structure itself which has very large forces - or possibly an issue with non-SCF forces.  I'll check this, and also put out a bug fix for the g0 error.

**Reply by davidbowler** [reply]
Yes - there is definitely an issue with non-SCF forces! Running this calculation with SCF turned on converges quickly.  I'll try to characterise this separately.

**Comment by davidbowler** [comment]
I'm going to change the name of this discussion to "Indexing errors during relaxation".  The original title was:

Index '0' of dimension 1 of array 'halo_rem%ndimj' below lower bound of 1

but this error message seems to be less common than "index_trans: i_h2d wrong".  In both cases, the problem seems to be fixed by correctly defining the variable `g0` in the `lbfgs` subroutine.

**Comment by lionelalexandre** [comment]
Thank you very much Marius and Dave. Problem is fixed.
